### PR TITLE
Migrate lib/src/crypto_subtle.dart to use `dart:js_interop`

### LINKED ISSUE #83: Migrate lib/src/crypto_subtle.dart to use `dart:js_interop`
Using `dart:js_interop` or `package:web` instead of other libraries will make this wasm-ready.

Migrating https://github.com/google/webcrypto.dart/blob/master/lib/src/crypto_subtle.dart should be possible.

### PR DESCRIPTION
**Pull request by koji-1009**
Migrate web crypto api  code to `dart:js_interop`.

I have added `JS` to class names, like `JSWindow`, to distinguish it from classes that are not related to `dart:js_interop`. But if this is unnecessary, please point it out in review comments.

close #83

### CODE REVIEW

File: lib/src/testing/utils/testrunner.dart [hunk 1]
```diff
@@ -809,7 +809,7 @@ void _runTests<PrivateKey, PublicKey>(
       signature = c.signature;
     } else {
       test('create signature', () async {
-        signature = await r._signBytes!(
+        signature = await r._signBytes(
```
Comment (jonasfj) [inline lib/src/testing/utils/testrunner.dart:812]: Why these changes? Is it perhaps because we updated the language version when we bumped the SDK constraint? (Just guessing that maybe this had an impact on inference for nullability)

If so, can we consider bumping the SDK constraint in a differnet PR?

File: lib/src/crypto_subtle.dart [hunk 2]
```diff
@@ -73,22 +42,142 @@ Uint8List bigIntToUint8ListBigInteger(BigInt integer) {
...
+  /// https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/deriveBits
+  external JSPromise<JSArrayBuffer> deriveBits(
+    JSAny algorithm,
+    JSCryptoKey baseKey,
+    int length,
+  );
+
+  /// https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/generateKey
+  @JS('generateKey')
+  external JSPromise<JSCryptoKey> generateCreptoKey(
```
Comment (jonasfj) [inline lib/src/crypto_subtle.dart:116]:
```suggestion
  external JSPromise<JSCryptoKey> generateCryptoKey(
```

typo?

File: lib/src/crypto_subtle.dart [hunk 3]
```diff
@@ -73,22 +42,142 @@ Uint8List bigIntToUint8ListBigInteger(BigInt integer) {
...
+    JSAny algorithm,
+    bool extractable,
+    JSArray<JSString> keyUsages,
+  );
+
+  /// https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/generateKey
+  @JS('generateKey')
+  external JSPromise<JSCryptoKeyPair> generateCreptoKeyPair(
```
Comment (jonasfj) [inline lib/src/crypto_subtle.dart:124]: typo?

File: lib/src/crypto_subtle.dart [hunk 4]
```diff
@@ -73,22 +42,142 @@ Uint8List bigIntToUint8ListBigInteger(BigInt integer) {
...
   ///  * `'private'`
   ///  * `'public'`
   ///  * `'secret'`
-  String get type => getProperty(_jsObj, 'type');
+  /// https://developer.mozilla.org/en-US/docs/Web/API/CryptoKey/type
+  external String get type;
 
   /// True, if this key can be extracted.
-  bool get extractable => getProperty(_jsObj, 'extractable');
+
```
Comment (jonasfj) [inline lib/src/crypto_subtle.dart:178]:
```suggestion
///
```
missing empty documentation comment line?

File: lib/src/impl_js/impl_js.random.dart [hunk 5]
```diff
@@ -17,7 +17,7 @@ part of impl_js;
 void fillRandomBytes(TypedData destination) {
   try {
     subtle.getRandomValues(destination);
-  } on subtle.DomException catch (e) {
+  } on subtle.JSDomException catch (e) {
```
Comment (jonasfj) [inline lib/src/impl_js/impl_js.random.dart:20]: Does this work?

File: lib/src/impl_js/impl_js.utils.dart [hunk 6]
```diff
@@ -107,7 +107,7 @@ Future<T> _handleDomException<T>(
 }) async {
   try {
     return await fn();
-  } on subtle.DomException catch (e) {
+  } on subtle.JSDomException catch (e) {
```
Comment (jonasfj) [inline lib/src/impl_js/impl_js.utils.dart:110]: Will this work?

File: lib/src/crypto_subtle.dart [hunk 7]
```diff
@@ -245,242 +300,375 @@ class Algorithm {
...
+  external String get t;
 }
 
-/// Convert a [list] a Javascript object.
-Object _stringListToJsObj(List<String> list) => callConstructor(
-      _array,
-      list,
+TypedData getRandomValues(TypedData array) {
+  if (array is Uint8List) {
+    final result = window.crypto.getRandomValues(array.toJS) as JSUint8Array;
```
Comment (jonasfj) [inline lib/src/crypto_subtle.dart:333]: We could perhaps make this simpler by throwing away the return value.

We can just `return array`, since that's also what the API returns:
https://developer.mozilla.org/en-US/docs/Web/API/Crypto/getRandomValues#return_value

File: lib/src/impl_js/impl_js.random.dart (hunk 5 above)
Comment (jonasfj) [reply lib/src/impl_js/impl_js.random.dart:20]: Can we make some unit tests for `impl_js`.

Like:
```js
crypto.getRandomValues(new Uint8Array(1000000))
// throws:
// Uncaught DOMException: Failed to execute 'getRandomValues' on 'Crypto': The ArrayBufferView's byte length (1000000) exceeds the number of bytes of entropy available via this API (65536).
```

at-least in chrome.


-----------------


Are there scenarios where we could get `TypeError` or other such things when calling into JS, and would handle them correctly?

**Review by jonasfj — COMMENTED** [review]
Mostly just nits, a few dumb questions about exception handling.

But overall this looks good.

File: lib/src/testing/utils/testrunner.dart (hunk 1 above)
Comment (koji-1009) [reply lib/src/testing/utils/testrunner.dart:812]: As you may have guessed, this is what lint pointed out because of the newer version of dart. The following PR separates the Flutter and Dart version upgrades and `dart fix --apply`.

https://github.com/google/webcrypto.dart/pull/87

File: lib/src/crypto_subtle.dart (hunk 3 above)
Comment (koji-1009) [reply lib/src/crypto_subtle.dart:124]: Sorry, typo, I'll fix the two places together.

File: lib/src/crypto_subtle.dart (hunk 7 above)
Comment (koji-1009) [reply lib/src/crypto_subtle.dart:333]: Thank you! 
I will give it a try!

File: lib/src/crypto_subtle.dart (hunk 2 above)
Comment (koji-1009) [reply lib/src/crypto_subtle.dart]: It's typo. I fixed. [e9c8900](https://github.com/google/webcrypto.dart/pull/86/commits/e9c890031c926e235bed935a9bf590a3bde4e7e5)

File: lib/src/crypto_subtle.dart (hunk 3 above)
Comment (koji-1009) [reply lib/src/crypto_subtle.dart]: It's typo. I fixed. [e9c8900](https://github.com/google/webcrypto.dart/pull/86/commits/e9c890031c926e235bed935a9bf590a3bde4e7e5)

File: lib/src/crypto_subtle.dart (hunk 4 above)
Comment (koji-1009) [reply lib/src/crypto_subtle.dart]: I added empty lines. [55af27e](https://github.com/google/webcrypto.dart/pull/86/commits/55af27e9403b77c5437f947ee9158f7b7805838a)

File: lib/src/impl_js/impl_js.random.dart (hunk 5 above)
Comment (koji-1009) [reply lib/src/impl_js/impl_js.random.dart:20]: I added a test file for each browser since the error is expressed differently in chrome and firefox. And, I was also able to add a scenario where DOMException is try-catch (a bit weird).
Does this look good?

* [chrome](https://github.com/google/webcrypto.dart/pull/86/files#diff-627e300589c395ca34cb9c192df9b8ea491ed4b16f67cdeaa9b90cd50a4cd441)
* [firefox](https://github.com/google/webcrypto.dart/pull/86/files#diff-8f2af7a8565ec85972baa2d7444e5765368cc42f8905f46df6cac4c6ab2b942d)

ref https://pub.dev/packages/test#restricting-tests-to-certain-platforms

File: lib/src/impl_js/impl_js.utils.dart (hunk 6 above)
Comment (koji-1009) [reply lib/src/impl_js/impl_js.utils.dart:112]: Strange as it may seem, try-catch is possible..

File: lib/src/impl_js/impl_js.random.dart (hunk 5 above)
Comment (jonasfj) [reply lib/src/impl_js/impl_js.random.dart:20]: Hmm, I guess this works :rocket:

I'm just getting used to that everything returned or thrown by these APIs is in fact a `JSObject`. And any instance of `JSObject` is also an instance of `subtle.JSDomException`.

So any JS exception thrown will be a `subtle.JSDomException`. But that's probably fine, since we always read the `.name` property in order to decide what to do.

File: test/browser_chrome_test.dart [hunk 8]
```diff
@@ -0,0 +1,186 @@
+@TestOn('chrome')
```
Comment (jonasfj) [inline test/browser_chrome_test.dart:1]: I imagine that the two files for Firefox and Chrome are quite similar.

Can we make this a single file.

Perhaps we can make `detectedRuntime` to handle different browsers differently.

If we only have a single file it's easier to spot the difference.

**Review by jonasfj — APPROVED** [review]
I think this is good enough to ship!

r+

If you want to tweak the test things, feel free. Otherwise, I'll land this, and be happy to accept PRs improving tests.

Writing tests for `crypto_subtle.dart` actually makes sense. But it's a non-trivial amount of work, and certainly not something we should block this PR on.

**Comment by jonasfj** [comment]
I'll get this out in the next release, feel free to file a PR preparing a release (bump version and update changelog), if you want it sooner.

**Comment by koji-1009** [comment]
@jonasfj 
Thanks for the review and advice!
I will work on improving the test over the weekend.
